File: passdrill.go
// passdrill: typing drills for practicing passphrases

package main

import (
	"bufio"
	"bytes"
	"crypto/rand"
	"crypto/sha512"
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"os"
	"strings"

	"github.com/howeyc/gopass"
	"golang.org/x/crypto/pbkdf2"
)

const hashFilename = "passdrill.hash"
const help = "Use -s to save passphrase hash for practice."

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func input(msg string) string {
	response := ""
	fmt.Print(msg)
	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		response = scanner.Text()
	}
	check(scanner.Err())
	return response
}

func prompt() string {
	fmt.Println("WARNING: the passphrase will be shown so that you can check it!")
	confirmed := ""
	passwd := ""
	for confirmed != "y" {
		passwd = input("Type passphrase to hash (it will be echoed): ")
		if passwd == "" || passwd == "q" {
			fmt.Println("ERROR: the passphrase cannot be empty or 'q'.")
			continue
		}
		fmt.Println("Passphrase to be hashed ->", passwd)
		confirmed = strings.ToLower(input("Confirm (y/n): "))
	}
	return passwd
}

func myPbkdf2(salt, content []byte) []byte {
	algorithm := sha512.New
	rounds := 200000
	keyLen := 64
	return pbkdf2.Key(content, salt, rounds, keyLen, algorithm)
}

func computeHash(keyFunc string, salt []byte, text string) []byte {
	if keyFunc == "pbkdf2" {
		return myPbkdf2(salt, []byte(text))
	}
	panic("Unknown key function " + keyFunc)
}

func buildHash(keyFunc, text string) []byte {
	salt := make([]byte, 32)
	_, err := rand.Read(salt)
	check(err)
	octets := computeHash(keyFunc, salt, text)
	headerStr := keyFunc + ":" + base64.StdEncoding.EncodeToString(salt) +
		":" + base64.StdEncoding.EncodeToString(octets[:])
	return []byte(headerStr)
}

func saveHash(args []string) {
	if len(os.Args) > 2 || os.Args[1] != "-s" {
		fmt.Println("ERROR: invalid argument.", help)
		os.Exit(1)
	}
	wrappedHash := buildHash("pbkdf2", prompt())
	err := ioutil.WriteFile(hashFilename, wrappedHash, 0600)
	check(err)
	fmt.Println("Passphrase hash saved to", hashFilename)
}

func unwrapHash(wrappedHash []byte) (string, []byte, []byte) {
	fields := strings.Split(string(wrappedHash), ":")
	keyFunc := fields[0]
	salt, err := base64.StdEncoding.DecodeString(fields[1])
	check(err)
	passwdHash, err := base64.StdEncoding.DecodeString(fields[2])
	check(err)
	return keyFunc, salt, passwdHash
}

func practice() {
	wrappedHash, err := ioutil.ReadFile(hashFilename)
	if os.IsNotExist(err) {
		fmt.Println("ERROR: passphrase hash file not found.", help)
		os.Exit(1)
	}
	check(err)
	keyFunc, salt, passwdHash := unwrapHash(wrappedHash)
	fmt.Println("Type q to end practice.")
	turn := 0
	correct := 0
	for {
		turn++
		fmt.Printf("%d:", turn)
		octets, err := gopass.GetPasswd()
		check(err)
		response := string(octets)
		if response == "" {
			fmt.Println("Type q to quit.")
			turn-- // don't count this response
			continue
		} else if response == "q" {
			turn-- // don't count this response
			break
		}
		answer := "wrong"
		if bytes.Compare(computeHash(keyFunc, salt, response), passwdHash) == 0 {
			correct++
			answer = "OK"
		}
		fmt.Printf("  %s\thits=%d\tmisses=%d\n", answer, correct, turn-correct)
	}
	if turn > 0 {
		pct := float64(correct) / float64(turn) * 100
		fmt.Printf("\n%d exercises. %0.1f%% correct.\n", turn, pct)
	}
}

func main() {
	if len(os.Args) > 1 {
		saveHash(os.Args)
	} else {
		practice()
	}
}

File: passdrill.py
#!/usr/bin/env python3

"""passdrill: typing drills for practicing passphrases
"""

import sys
import hashlib
import base64
import getpass
import os

HASH_FILENAME = 'passdrill.hash'
HELP = 'Use -s to save passphrase hash for practice.'


def prompt():
    print('WARNING: the passphrase will be shown so that you can check it!')
    confirmed = ''
    while confirmed != 'y':
        passwd = input('Type passphrase to hash (it will be echoed): ')
        if passwd == '' or passwd == 'q':
            print('ERROR: the passphrase cannot be empty or "q".')
            continue
        print(f'Passphrase to be hashed -> {passwd}')
        confirmed = input('Confirm (y/n): ').lower()
    return passwd


def pbkdf2(salt, octets):
    algorithm = 'sha512'
    rounds = 100_000
    return hashlib.pbkdf2_hmac(algorithm, octets, salt, rounds)


def compute_hash(key_func, salt, text):
    octets = text.encode('utf-8')
    if key_func == 'pbkdf2':
        return pbkdf2(salt, octets)
    else:
        raise ValueError('Unknown key function ' + repr(key_func))


def build_hash(key_func, text):
    salt = os.urandom(32)
    octets = compute_hash(key_func, salt, text)
    header = key_func.encode('utf-8') + b':' + base64.b64encode(salt)
    return header + b':' + base64.b64encode(octets)


def save_hash(argv):
    if len(argv) > 2 or argv[1] != '-s':
        print('ERROR: invalid argument.', HELP)
        sys.exit(1)
    wrapped_hash = build_hash('pbkdf2', prompt())
    with open(HASH_FILENAME, 'wb') as fp:
        fp.write(wrapped_hash)
    print(f'Passphrase hash saved to', HASH_FILENAME)


def unwrap_hash(wrapped_hash):
    key_func, salt, passwd_hash = wrapped_hash.split(b':')
    return (key_func.decode('utf-8'), base64.b64decode(salt),
        base64.b64decode(passwd_hash))


def practice():
    try:
        with open(HASH_FILENAME, 'rb') as fp:
            wrapped_hash = fp.read()
    except FileNotFoundError:
        print('ERROR: passphrase hash file not found.', HELP)
        sys.exit(2)
    key_func, salt, passwd_hash = unwrap_hash(wrapped_hash)
    print('Type q to end practice.')
    turn = 0
    correct = 0
    while True:
        turn += 1
        response = getpass.getpass(f'{turn}:')
        if response == '':
            print('Type q to quit.')
            turn -= 1  # don't count this response
            continue
        elif response == 'q':
            turn -= 1  # don't count this response
            break
        if compute_hash(key_func, salt, response) == passwd_hash:
            correct += 1
            answer = 'OK'
        else:
            answer = 'wrong'
        print(f'  {answer}\thits={correct}\tmisses={turn-correct}')

    if turn:
        pct = correct / turn * 100
        print(f'\n{turn} exercises. {pct:0.1f}% correct.')


if __name__ == '__main__':
    if len(sys.argv) > 1:
        save_hash(sys.argv)
    else:
        practice()
